[WebDriver] Add flag to allow replacing a stale session

lauromoura opened the pull request:
#### 16ef019e907d305c91ca4e90b22273380fd5482a
<pre>
[WebDriver] Add flag to allow replacing a stale session
<a href="https://bugs.webkit.org/show_bug.cgi?id=287540">https://bugs.webkit.org/show_bug.cgi?id=287540</a>

Reviewed by Carlos Garcia Campos.

Ideally, the client code should take care of properly cleaning up
existing sessions. But if it happens to fail to do so, the driver might
get stuck with a stale session, preventing the creation of new ones,
given our current policy of having at most one session active.

This commit adds an option to avoid this by with a new flag that
destroys the currently active session when requested to create new ones.

Also, drive-by safer-cpp fix on nearby code, replacing printf with
SAFE_PRINT.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run):
(WebDriver::WebDriverService::newSession):
(WebDriver::WebDriverService::deleteSession):
* Source/WebDriver/WebDriverService.h:

Canonical link: <a href="https://commits.webkit.org/290489@main">https://commits.webkit.org/290489@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c32801635c2c8ec0d7f5bb25b12cb3d804c8d79d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45106 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95186 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40961 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18027 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69431 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27035 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93186 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49792 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36189 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40092 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77796 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37243 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97011 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12776 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78430 "layout-tests (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17630 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77635 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19171 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22090 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20693 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10632 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17383 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22709 "") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17124 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20576 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18908 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebDriver/WebDriverService.h:
```diff
@@ -183,6 +183,8 @@ class WebDriverService final : public HTTPRequestHandler
     String m_targetAddress;
     uint16_t m_targetPort { 0 };
+
+    bool m_replaceOnNewSession;
```
```suggestion
    bool m_replaceOnNewSession { false };
```

lauromoura committed:
[WebDriver] Add flag to allow replacing a stale session
https://bugs.webkit.org/show_bug.cgi?id=287540

Reviewed by Carlos Garcia Campos.

Ideally, the client code should take care of properly cleaning up
existing sessions. But if it happens to fail to do so, the driver might
get stuck with a stale session, preventing the creation of new ones,
given our current policy of having at most one session active.

This commit adds an option to avoid this by with a new flag that
destroys the currently active session when requested to create new ones.

Also, drive-by safer-cpp fix on nearby code, replacing printf with
SAFE_PRINT.

* Source/WebDriver/WebDriverService.cpp:
(WebDriver::printUsageStatement):
(WebDriver::WebDriverService::run):
(WebDriver::WebDriverService::newSession):
(WebDriver::WebDriverService::deleteSession):
* Source/WebDriver/WebDriverService.h:

Canonical link: https://commits.webkit.org/290489@main
